rust/extractor/src/translate.rs

```diff
@@ -80,6 +156,671 @@ impl CrateTranslator<'_> {
+                    id: TrapId::Star,
+                    location,
+                    inner,
+                })
+            }
+            ra_ap_hir_def::hir::Pat::ConstBlock(expr) => {
+                let expr = self.emit_expr(*expr, body, source_map);
+                self.trap.emit(generated::ConstBlockPat {
+                    id: TrapId::Star,
+                    location,
+                    expr,
+                })
+            }
+        }
+    }
+    fn emit_type_ref(
+        &mut self,
+        type_ref: &TypeRef, //TODO
+    ) -> trap::Label {
+        let location: Option<trap::Label> = None; //TODO: ?
+        self.trap.emit(generated::TypeRef {
+            id: TrapId::Star,
+            location,
+        })
+    }
+    fn emit_match_arm(
+        &mut self,
+        arm: &MatchArm,
+        body: &Body,
+        source_map: &BodySourceMap,
+    ) -> trap::Label {
+        let location: Option<trap::Label> = self.emit_location_pat(arm.pat, source_map);
+        let pat = self.emit_pat(arm.pat, body, source_map);
+        let guard = arm.guard.map(|g| self.emit_expr(g, body, source_map));
+        let expr = self.emit_expr(arm.expr, body, source_map);
+        self.trap.emit(generated::MatchArm {
+            id: TrapId::Star,
+            location,
+            pat,
+            guard,
+            expr,
+        })
+    }
+    fn emit_stmt(
+        &mut self,
+        stmt: &Statement,
+        body: &Body,
+        source_map: &BodySourceMap,
+    ) -> trap::Label {
+        match stmt {
+            Statement::Let {
+                pat,
+                type_ref,
+                initializer,
+                else_branch,
+            } => {
+                // TODO: find a way to get the location of the entire statement
+                let location = self.emit_location_pat(*pat, source_map);
+                let pat = self.emit_pat(*pat, body, source_map);
+                let type_ref = type_ref
+                    .as_ref()
+                    .map(|type_ref| self.emit_type_ref(type_ref));
+                let initializer =
+                    initializer.map(|initializer| self.emit_expr(initializer, body, source_map));
+                let else_ = else_branch.map(|else_| self.emit_expr(else_, body, source_map));
+
+                self.trap.emit(generated::IfLet {
+                    id: TrapId::Star,
+                    location,
+                    pat,
+                    type_ref,
+                    initializer,
+                    else_,
+                })
+            }
+            Statement::Expr { expr, has_semi } => {
+                let location = self.emit_location_expr(*expr, source_map);
+                let expr = self.emit_expr(*expr, body, source_map);
+                self.trap.emit(generated::ExprStmt {
+                    id: TrapId::Star,
+                    location,
+                    expr,
+                    has_semi: *has_semi,
```

Comment:
I'm guessing this means semicolon? If yes, it might be worth spelling it out entirely